### PR TITLE
Fractional Differening

### PR DESCRIPTION
**Pull request by ashuWEMBA**
# Description

Added Fractional Differentiation code with unit-tests in this release.  Made the code compliant with coding guidelines (comments, variable names and function names).

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

We ran a unit test for each function with test data.  The test data are sample dollar bars included as part of the package.  

**Test Configuration**:
* Operating system: Windows 7, 64-bit
* IDE used: PyCharm

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: mlfinlab/features/fracdiff.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+        for iloc in range(skip, series_f.shape[0]):
+            loc = series_f.index[iloc]
+
+            # At this point all entries are non-NAs so no need for the following check
+            # if np.isfinite(series.loc[loc, name]):
+            output_df_[loc] = np.dot(weights[-(iloc + 1):, :].T, series_f.loc[:loc])[0, 0]
+
+        output_df[name] = output_df_.copy(deep=True)
+    output_df = pd.concat(output_df, axis=1)
```
Comment (Jackal08) [inline mlfinlab/features/fracdiff.py]: What is this line doing? Transposing? We are concatenating a single series.

File: mlfinlab/features/fracdiff.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    The discussion of positive and negative d is similar to that in get_weights
+    (see the function get_weights)
+
+    :param diff_amt: (float) differencing amount
+    :param thresh: (float) threshold for minimum weight
+    :param lim: (int) maximum length of the weight vector
+    :return: (ndarray) weight vector
+    """
+
+    weights, k = [1.], 1
```
Comment (Jackal08) [inline mlfinlab/features/fracdiff.py]: Please, can we add comments to this code to improve readability?

File: mlfinlab/tests/test_fracdiff.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        self.path = project_path + '/test_data/dollar_bar_sample.csv'
+        self.data = pd.read_csv(self.path, index_col='date_time')
+        self.data.index = pd.to_datetime(self.data.index)
+
+    def test_get_weights(self):
+        """
+        get_weights as implemented here matches the code in the book (Snippet 5.1).
+
+        We test:
+        1. if the first element of the weights vector is 1.0
```
Comment (Jackal08) [inline mlfinlab/tests/test_fracdiff.py]: I like this docstring structure for unit tests. I will start to do it as well.

File: mlfinlab/tests/test_fracdiff.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        # Size matches
+        self.assertTrue(weights.shape[0] == number_ele)
+
+    def test_get_weights_ffd(self):
+        """
+        get_weights_ffd as implemented here matches the code in the book (Snippet 5.2).
+        We test:
+        1. if the first element of the weights vector is 1.0
+        2. The length of the weights vector is equal to 12
+
```
Comment (Jackal08) [inline mlfinlab/tests/test_fracdiff.py]: Remove white space

File: mlfinlab/tests/test_fracdiff.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        1. Length of the output is the same as the length of the input
+        2. First element is NaN
+        """
+        data_series = self.data['close'].to_frame()
+
+        for diff_amt in np.arange(0.1, 1, 0.1):
+            fd_series = fracdiff.frac_diff_ffd(data_series, diff_amt=diff_amt)
+            self.assertTrue(fd_series.shape[0] == len(data_series))
+            self.assertTrue(isinstance(fd_series['close'][0], np.float64) and math.isnan(fd_series['close'][0]))
+            #  self.assertTrue(type(fd_series['close'][0]) is np.float64 and math.isnan(fd_series['close'][0]))
```
Comment (Jackal08) [inline mlfinlab/tests/test_fracdiff.py]: Please remove this line

File: mlfinlab/tests/test_fracdiff.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
+"""
```
Comment (Jackal08) [inline mlfinlab/tests/test_fracdiff.py]: I am happy with this if you feel that the unit tests here cover enough edge cases.

File: mlfinlab/features/fracdiff.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
+"""
```
[the remark this answers is not in the document]
Comment (ashuWEMBA) [reply mlfinlab/features/fracdiff.py]: Good point ... I actually like that structure and if you two agree we can use that ... it is cleaner in my opinion and more forward looking.

File: mlfinlab/features/fracdiff.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    number of elements.  And that translates into a set whose elements can be
+    selected more than once or as many times as one chooses (multisets with
+    unbounded multiplicity) - see http://faculty.uml.edu/jpropp/msri-up12.pdf.
+
+    :param diff_amt: (float) differencing amount
+    :param size: (int) length of the series
+    :return: (ndarray) weight vector
+    """
+
+    weights = [1.]
```
[the remark this answers is not in the document]
Comment (ashuWEMBA) [reply mlfinlab/features/fracdiff.py]: Good idea

File: mlfinlab/features/fracdiff.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+        weights_ = -weights[-1] / k * (diff_amt - k + 1)
+        weights.append(weights_)
+
+    weights = np.array(weights[::-1]).reshape(-1, 1)
+    return weights
+
+
+def frac_diff(series, diff_amt, thresh=0.01):
+    """
+    Source: Chapter 5, AFML (section 5.5);
```
[the remark this answers is not in the document]
Comment (ashuWEMBA) [reply mlfinlab/features/fracdiff.py]: ok

File: mlfinlab/features/fracdiff.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    # 3. Apply weights to values
+    output_df = {}
+    for name in series.columns:
+        series_f = series[[name]].fillna(method='ffill').dropna()
+        output_df_ = pd.Series(index=series.index)
+
+        for iloc in range(skip, series_f.shape[0]):
+            loc = series_f.index[iloc]
+
+            # At this point all entries are non-NAs so no need for the following check
```
[the remark this answers is not in the document]
Comment (ashuWEMBA) [reply mlfinlab/features/fracdiff.py]: That is why I let that in.

File: mlfinlab/features/fracdiff.py (hunk 1 above)
Comment (ashuWEMBA) [reply mlfinlab/features/fracdiff.py]: Suppose the input data belonged to several stocks and we wanted to compute fractionally differenced series for all of them.  To meet that need, output_df was created as a dict with the name of the stock as the key.  The loop computes the frac-diff for each stock and adds to the dictionary.   This line converts that dictionary into a data frame

File: mlfinlab/features/fracdiff.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Jackal08) [reply mlfinlab/features/fracdiff.py]: I am still a bit confused as to which structure to use. From what I understood in the python textbook is that this structure is more pythonic but I may be wrong. I think this structure is fine for now but we need to read those python chapters and have a discussion around this.
